xc/device.ml

```diff
@@ -1963,6 +1974,96 @@ module Dm = struct
         List.iter (write_local_domain "data/") ["updated"]
       | _ -> ()
 
+  (* Create two hash tables to store <domid, qmp connction> and <qmp connection, domid> *)
+  let domid_to_connection_tbl = Hashtbl.create 128
+  let connection_to_domid_tbl = Hashtbl.create 128
+
+  let qmp_event_handle domid qmp_event =
+    (* This function will be extended to handle qmp events *)
+    debug "Get QMP event, domain-%d: %s" domid qmp_event.event
+ 
+  let qmp_event_thread () =
+    let epfd = Epoll.create1 [Epoll.EPOLL_CLOEXEC] in
+    Epoll.ctl epfd Epoll.EPOLL_CTL_ADD qmp_event_pipe_r [Epoll.EPOLLIN];
+
+    let remove domid =
+      if Hashtbl.mem domid_to_connection_tbl domid
+      then begin
+        try
+          let c = Hashtbl.find domid_to_connection_tbl domid in
+          Hashtbl.remove domid_to_connection_tbl domid;
+          Hashtbl.remove connection_to_domid_tbl c;
+          Epoll.ctl epfd Epoll.EPOLL_CTL_DEL (Qmp_protocol.to_fd c) [Epoll.EPOLLIN];
+          Qmp_protocol.close c
+        with
+        | Unix.Unix_error(Unix.EBADF, "close", p) -> raise(Internal_error (Printf.sprintf "Failed to close qmp socket on domain-%d: %s" domid p))
+        | e -> raise(Internal_error (Printf.sprintf "Get error trying remove qmp on domain-%d: %s" domid (Printexc.to_string e)))
+      end
+    in
+
+    let add domid =
+      if not(Hashtbl.mem domid_to_connection_tbl domid)
+      then begin
+        let path = Printf.sprintf "/var/run/xen/qmp-event-libxl-%d" domid in
+        try
+          let c = Qmp_protocol.connect path in
+          Qmp_protocol.negotiate c;
+          Hashtbl.replace domid_to_connection_tbl domid c;
+          Hashtbl.replace connection_to_domid_tbl c domid;
+          Epoll.ctl epfd Epoll.EPOLL_CTL_ADD (Qmp_protocol.to_fd c) [Epoll.EPOLLIN];
+          debug "QMP domain-%d: negotiation complete" domid
+        with e ->
+          debug "QMP domain-%d: negotiation failed (%s): removing socket" domid (Printexc.to_string e);
+          remove domid
+      end
+    in
+
+    while true do
```

Comment:
I think we need a try-with protecting this loop; we shouldn't ever leave this loop while xenopsd is running.

Reply:
Added, thanks.